### PR TITLE
Refactor of FirebaseDB

### PR DESCRIPTION
**Pull request by LucaCarroz**
Small refactor of FirebaseDB that now uses Firebase's storage to store images instead of realtime database.
Also refactored tests a bit and added some edge case tests.
Added some comments

### CODE REVIEW

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt
Comment (david-schroeter) [inline app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:1]: Overall this file looks really good, good idea to have use an object to make this class a singleton !

But I am just not entirely sure we should keep that much commented code in the file. That is true that it could be useful code that we might use later, but I think it would be better to maybe store this code elsewhere, and if we use it some day we can take it rather than having commented code forever in this file if we end up not using it.

What do you think?

**Review by david-schroeter — CHANGES_REQUESTED** [review]
Very good PR. Nice improvement to use the storage now to store picture that is a lot better. 

The only important point I think is about the tests, we should probably load the data in the emulator when launching it rather than adding it in the test setup, as it prevent the correct functioning of the emulator.

File: app/src/androidTest/java/com/github/sdp_begreen/begreen/activities/DatabaseActivityTest.kt [hunk 1]
```diff
@@ -8,24 +8,30 @@ import androidx.test.espresso.matcher.ViewMatchers.*
...
-        Firebase.database.useEmulator("10.0.2.2",9000)
+    companion object {
+        @BeforeClass @JvmStatic fun setup() {
+            try {
+                Firebase.database.useEmulator("10.0.2.2", 9000)
+                Firebase.storage.useEmulator("10.0.2.2", 9199)
+                Firebase.auth.useEmulator("10.0.2.2", 9099)
+            } catch (_:java.lang.IllegalStateException){}
+
+            Firebase.database.reference.child("123").setValue("abc")
```
[the remark this answers is not in the document]
Comment (LucaCarroz) [reply app/src/androidTest/java/com/github/sdp_begreen/begreen/activities/DatabaseActivityTest.kt:33]: True! Mysteriously, this still worked yesterday but it seems to cause a problem now, I'll get rid of it to prevent any bugs, thanks for pointing that out!

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt
Comment (LucaCarroz) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:1]: Thats true that it makes the code a bit unreadable, I'll keep a copy of those ideas locally and remove them from the git

**Review by david-schroeter — APPROVED** [review]
Thanks for the changes, it looks all good to me now!

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+     */
+    operator fun set(key: String, value: String) {
+        if (key.isNotEmpty()) databaseReference.child(key).setValue(value)
+    }
+
+    /**
+     * Store the given [image] under the ID [userId]
+     * and returns the imageId set for that image in the database
+     * Returns null if the image can't be stored
+     */
```
Comment (SolalPirelli) [inline app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:93]: This comment looks orphaned, as if it was applying to a method that no longer exists

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+     * Retrieves the image associated with the given [userId] and [imageId] from the database
+     *
+     * @param imageId the ID of the image we want to retrieve from the database
+     * @param userId the ID of the user where we should find the image
+     * @return the image, or null if no image was found
+     * @throws StorageException if the image could not get retrieved
+     * @throws DatabaseException if the metadata could not get retrieved
+     *
+     */
+    @SuppressLint("RestrictedApi")
```
Comment (SolalPirelli) [inline app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:139]: This does not look like a good idea? https://stackoverflow.com/a/64818775

If there is a good reason to do it, it's exactly what comments are for :)

File: storage.rules [hunk 4]
```diff
@@ -0,0 +1,9 @@
+rules_version = '2';
+service firebase.storage {
+  match /b/{bucket}/o {
+    match /{allPaths=**} {
+      allow read, write: if
+          request.time < timestamp.date(2023, 4, 13);
```
Comment (SolalPirelli) [inline storage.rules:6]: 🤔

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt (hunk 3 above)
Comment (LucaCarroz) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:139]: Yes indeed, after discussions David changed it in his PR just after #30

File: app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt (hunk 2 above)
Comment (LucaCarroz) [reply app/src/main/java/com/github/sdp_begreen/begreen/firebase/FirebaseDB.kt:93]: That's probably the case. David or me must have spotted in when working on `FirebaseDB` in our later tasks, because it is not present anymore. Thanks though!
